src/Modules/OrchardCore.Commerce/Endpoints/Services/ShoppingCartService.cs

```diff
@@ -0,0 +1,145 @@
+using Lombiq.HelpfulLibraries.AspNetCore.Exceptions;
+using Lombiq.HelpfulLibraries.OrchardCore.Workflow;
+using OrchardCore.Commerce.Abstractions;
+using OrchardCore.Commerce.Abstractions.Abstractions;
+using OrchardCore.Commerce.Abstractions.Models;
+using OrchardCore.Commerce.Activities;
+using OrchardCore.Commerce.Endpoints.Extensions;
+using OrchardCore.Commerce.Inventory.Models;
+using OrchardCore.Commerce.ViewModels;
+using OrchardCore.ContentManagement;
+using OrchardCore.Workflows.Services;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Commerce.Endpoints;
+public class ShoppingCartService : IShoppingCartService
+{
+    private readonly IShoppingCartHelpers _shoppingCartHelpers;
+    private readonly IShoppingCartPersistence _shoppingCartPersistence;
+    private readonly IShoppingCartSerializer _shoppingCartSerializer;
+    private readonly IEnumerable<IWorkflowManager> _workflowManagers;
+    private readonly IEnumerable<IShoppingCartEvents> _shoppingCartEvents;
+    private readonly IProductService _productService;
+    public ShoppingCartService(
+        IShoppingCartHelpers shoppingCartHelpers,
+        IShoppingCartPersistence shoppingCartPersistence,
+        IShoppingCartSerializer shoppingCartSerializer,
+        IEnumerable<IWorkflowManager> workflowManagers,
+        IEnumerable<IShoppingCartEvents> shoppingCartEvents,
+        IProductService productService)
+    {
+        _shoppingCartHelpers = shoppingCartHelpers;
+        _shoppingCartPersistence = shoppingCartPersistence;
+        _shoppingCartSerializer = shoppingCartSerializer;
+        _workflowManagers = workflowManagers;
+        _shoppingCartEvents = shoppingCartEvents;
+        _productService = productService;
+    }
+
+    public async Task<string> RemoveLineAsync(ShoppingCartLineUpdateModel line, string shoppingCartId = null)
+    {
+        string errored = string.Empty;
+        try
+        {
+            var parsedLine = await _shoppingCartSerializer.ParseCartLineAsync(line);
+            var cart = await _shoppingCartPersistence.RetrieveAsync(shoppingCartId);
+            cart.RemoveItem(parsedLine);
+            await _shoppingCartPersistence.StoreAsync(cart, shoppingCartId);
+        }
+        catch (Exception ex)
+        {
+            errored = ex.Message;
```

Comment:
Don't forward all exception messages (only FrontendException), it's a security risk. Use a generic text like "An error has occurred.".